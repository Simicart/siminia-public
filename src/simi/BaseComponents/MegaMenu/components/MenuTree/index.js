export { default } from './menuTree';
export { default as SubCategories } from './subCategories';
