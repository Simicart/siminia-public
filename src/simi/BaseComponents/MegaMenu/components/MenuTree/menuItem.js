import React, { useCallback, useEffect, useState } from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { number, shape, string } from 'prop-types';
import defaultClasses from './menuItem.module.css';
import MenuLink from './menuLink';
import SubMenuContent from './SubMenuContent';
import { useAmMegaMenuContext } from '../../context';
import { isEnableMenuItem } from '../../utils';
import { useAppContext } from '@magento/peregrine/lib/context/app';

const PATCH_TO_IMG = '/media/amasty/megamenu/submenu_background_image/';

const MenuItem = props => {
    const { level, width, view, id, status } = props;
    const {
        isMobile,
        isShowIcons,
        config,
        toggleAction,
        toggleActionMobile,
        openCategoryIds
    } = useAmMegaMenuContext();

    const [, { closeDrawer }] = useAppContext();

    const [changePage, setChangePage] = useState(false);

    const isMainLevel = level === 1;

    const hoverHandler = useCallback(
        e => {
            if (isMainLevel && view === 'topMenu' && !changePage && !isMobile) {
                toggleAction(id);
            }
        },
        [isMainLevel, toggleAction, view, id, changePage, isMobile]
    );

    const clickHandler = useCallback(
        e => {
            if (isMainLevel && view === 'topMenu' && isMobile) {
                toggleActionMobile(id);
            }
        },
        [isMainLevel, toggleAction, view, id, changePage, isMobile]
    );

    const onNavigation = useCallback(() => {
        setChangePage(true);
        toggleAction(id);
        isMobile && closeDrawer();
    }, [id, isMobile]);

    useEffect(() => {
        if (changePage) {
            setTimeout(() => {
                setChangePage(false);
            }, 1000);
        }
    }, [changePage]);

    if (!isEnableMenuItem(status, isMobile)) {
        return null;
    }

    const classes = mergeClasses(defaultClasses, props.classes);
    const isOpen = openCategoryIds && openCategoryIds.has(id);
    const openClass = isOpen ? classes.open : '';
    const levelClass = isMainLevel ? classes.main : classes.child;

    // const widthClass = isMainLevel
    //   ? classes[widthTypes[width] || widthTypes[0]]
    //   : '';

    const rootClass = [classes.root, openClass, classes[view], levelClass].join(
        ' '
    );

    const {
        ammegamenu_color_submenu_text,
        ammegamenu_color_submenu_background_color,
        ammegamenu_color_color_template,
        ammegamenu_color_submenu_background_image,
        ammegamenu_color_main_menu_text_hover,
        ammegamenu_color_submenu_text_hover,
        ammegamenu_color_main_menu_background_hover
    } = config;

    const amStyle =
        isMainLevel && ammegamenu_color_color_template !== 'blank'
            ? {
                  '--am-mega-menu-submenu-text-color': ammegamenu_color_submenu_text,
                  '--am-mega-menu-submenu-text-hover-color': ammegamenu_color_submenu_text_hover,
                  '--am-mega-menu-submenu-bg-color': ammegamenu_color_submenu_background_color,
                  '--am-mega-menu-text-hover-color': ammegamenu_color_main_menu_text_hover,
                  '--am-mega-menu-item-bg-hover': ammegamenu_color_main_menu_background_hover,
                  '--am-mega-menu-submenu-bg-image':
                      view === 'topMenu' &&
                      ammegamenu_color_submenu_background_image
                          ? `url("${PATCH_TO_IMG +
                                ammegamenu_color_submenu_background_image}")`
                          : ammegamenu_color_submenu_background_color
              }
            : null;

    const menuItemNode = document.getElementById('menu-item-' + id);
    const menuItemContainerNode = document.getElementById(
        'menu-item-container-' + id
    );
    const menuTreeNode = document.getElementById('menu-tree');
    const style = { pointerEvents: 'none' };
    let subContainerMainClass = classes.subContainerMain;
    if (isOpen) {
        if (!isMobile) {
            if (amStyle && menuItemNode && menuTreeNode) {
                const menuTreeHeight = menuTreeNode.clientHeight;
                const offsetTop = menuItemNode.offsetTop;
                const positionMenuItem = offsetTop + menuItemNode.clientHeight;
                const percent = positionMenuItem / menuTreeHeight;
                if (percent > 0) {
                    style.top = `calc(${100 * percent}% - 2px)`;
                }
            }
        } else {
            if (menuItemContainerNode) {
                const menuHeight = menuItemContainerNode.clientHeight;
                style.height = menuHeight;
            }
        }
        style.pointerEvents = 'all';
        subContainerMainClass = classes.subContainerMainOpen;
    }

    return (
        <li
            id={'menu-item-' + id}
            className={rootClass}
            style={amStyle}
            onMouseEnter={hoverHandler}
            onMouseLeave={hoverHandler}
            onClick={clickHandler}
            onFocus={() => void 0}
        >
            <MenuLink
                {...props}
                toggleAction={toggleAction}
                isOpen={isOpen}
                onNavigate={onNavigation}
                isShowIcons={isShowIcons}
                isMainTitle={true}
            />
            <SubMenuContent
                {...props}
                menuItemId={id}
                isMobile={isMobile}
                isOpen={isOpen}
                isShowIcons={isShowIcons}
                hideIcon={true}
                style={style}
                onNavigation={onNavigation}
                classes={{
                    root: classes.subContainerRoot,
                    main: subContainerMainClass
                }}
            />
        </li>
    );
};

MenuItem.propTypes = {
    id: string,
    view: string,
    width: number,
    level: number,
    classes: shape({
        root: string
    })
};

MenuItem.defaultProps = {
    level: 1
};

export default MenuItem;
