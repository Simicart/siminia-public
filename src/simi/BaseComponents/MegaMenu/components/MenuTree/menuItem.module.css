:root {
  --am-mega-menu-submenu-text-color: var(--venia-global-color-text);
  --am-mega-menu-submenu-bg-color: var(--venia-global-color-background);
  --am-mega-menu-submenu-bg-image: var(--am-mega-menu-submenu-bg-color);
  --am-mega-menu-text-hover-color: var(--venia-global-color-teal);
  --am-mega-menu-submenu-text-hover-color: var(--venia-global-color-teal);
}

.root {
  position: relative;
}

.topMenu.main {
  padding: 0 0.5rem;
}

.topMenu.fullWidth {
  position: static;
}

.subContainer {
  display: none;
}

.topMenu.main:hover .subContainerMain:not(:empty) {
  display: block;
  z-index: 2;
}

.hamburger.main {
  align-items: center;
  border-bottom: 1px solid rgba(var(--am-mega-menu-text-color) / 0.2);
  margin: 0 1.25rem;
}

.subContainerRoot {
  display: none;
}

.subContainerMain {
  padding-bottom: 1rem;
  max-width: 100%;
}

.hamburger.open > .subContainerRoot:not(:empty) {
  display: block;
  padding-left: 1rem;
}

/*
 * Desktop specific styles
 */

@media (min-width: 1025px) {
  .subContainerMain {
    --venia-global-color-text: var(--am-mega-menu-submenu-text-color);
    position: absolute;
    padding: 2rem;
    border-radius: 1px;
    box-shadow: 0 1px 6px rgb(var(--venia-global-color-gray-900) / 0.2);
    max-width: initial;
    color: rgb(var(--venia-global-color-text));
    background-color: rgb(var(--am-mega-menu-submenu-bg-color));
    background-image: var(--am-mega-menu-submenu-bg-image);
    background-size: auto;
    background-position: center center;
    background-repeat: no-repeat;
    max-height: 70vh;
    overflow-y: auto;
  }

  .hamburger.main.open > .subContainerMain:not(:empty) {
    display: block;
    top: 0;
    left: 100%;
    width: max-content;
    margin-left: 1.25rem;
    max-width: calc(100vw - 24rem);
  }

  .fullWidth.hamburger.main.open > .subContainerMain {
    width: 100vw;
  }

  .fullWidth > .subContainerMain {
    left: 0;
    right: 0;
    width: 100vw;
    max-width: var(--venia-global-maxWidth);
    margin: 0 auto;
  }

  .autoWidth > .subContainerMain {
    width: max-content;
    max-width: 60vw;
  }
}
