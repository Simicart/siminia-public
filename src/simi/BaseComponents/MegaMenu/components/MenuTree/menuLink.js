import React, { useCallback } from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { number, shape, string } from 'prop-types';
import ToggleTrigger from './toggleTrigger';
import defaultClasses from './menuLink.module.css';
import { useLocation, Link } from 'react-router-dom';
import { useAppContext } from '@magento/peregrine/lib/context/app';

const LinkComponent = props => {
  const { to, children } = props;

  if (to && to.startsWith('http')) {
    return (
      <a {...props} href={to}>
        {children}
      </a>
    );
  }

  return (
    <Link {...props} to={`/${to}`}>
      {children}
    </Link>
  );
};

const MenuLink = props => {
  const {
    id,
    url,
    name,
    label,
    label_text_color,
    label_background_color,
    icon,
    subCategories,
    toggleAction,
    view,
    level,
    isOpen,
    isShowIcons,
    onNavigate
  } = props;

  const { pathname } = useLocation();
  const [, { closeDrawer }] = useAppContext();

  const handleNavigate = useCallback(() => {
    if (onNavigate) {
      return onNavigate();
    }

    return closeDrawer();
  }, [closeDrawer, onNavigate]);

  const classes = mergeClasses(defaultClasses, props.classes);

  const labelContent = label ? (
    <span
      style={{
        '--am-mega-menu-label-color': label_text_color,
        '--am-mega-menu-label-bg': label_background_color
      }}
      className={classes.label}
    >
      {label}
    </span>
  ) : null;

  const iconContent =
    icon && isShowIcons ? (
      <figure
        className={classes.icon}
        style={{
          WebkitMaskImage: `url("${icon}")`,
          maskImage: `url("${icon}")`,
          backgroundImage: icon.includes('jpg') ? `url("${icon}")` : undefined
        }}
      />
    ) : null;

  const toggleBtn =
    Array.isArray(subCategories) && subCategories.length ? (
      <ToggleTrigger
        classes={props.classes}
        action={() => toggleAction(id)}
        isOpen={isOpen}
      />
    ) : null;

  const levelClass = classes[`level${level}`];
  const activeClass =
    url && pathname.replace('.html', '').includes(url.replace('.html', ''))
      ? classes.active
      : '';
  const rootClass = [classes.root, levelClass, classes[view], activeClass].join(
    ' '
  );

  return (
    <div className={rootClass}>
      <LinkComponent
        className={classes.menuLink}
        onClick={handleNavigate}
        to={url}
      >
        {iconContent}

        <span className={classes.title}>{name}</span>

        {labelContent}
      </LinkComponent>

      {toggleBtn}
    </div>
  );
};

MenuLink.propTypes = {
  classes: shape({
    root: string
  }),
  url: string,
  name: string,
  label: string,
  label_text_color: string,
  label_background_color: string,
  icon: string,
  level: number
};

MenuLink.defaultProps = {
  level: 1
};

export default MenuLink;
