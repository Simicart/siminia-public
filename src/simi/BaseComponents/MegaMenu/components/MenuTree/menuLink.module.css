:root {
  --am-mega-menu-label-color: #000;
  --am-mega-menu-label-bg: #fff;
  --am-mega-menu-active-color: var(--venia-global-color-text);
  --am-mega-menu-item-bg-hover: var(--venia-global-color-background);
}

.root {
  display: flex;
  align-items: center;
  width: 100%;
  color: rgb(var(--venia-global-color-text));
}

.toggle {
  display: block;
  transform: rotate(90deg);
}

.toggle_open {
  composes: toggle;
  transform: rotate(-90deg);
}

.toggle button {
  width: 2rem;
  height: 3.5rem;
  outline: none;
}

.level1 {
  --venia-global-color-text: var(--am-mega-menu-text-color);
}

.hamburger:not(.level1) {
  --venia-global-color-text: var(--am-mega-menu-submenu-text-color);
}

.hamburger:not(.level1) .toggle svg {
  width: 1rem;
}

a.menuLink {
  height: 2.5rem;
  display: flex;
  align-items: center;
  white-space: nowrap;
  width: 100%;
  min-width: 0;
  text-decoration: none !important;
}

.hamburger.level1 > .menuLink {
  height: 3.5rem;
  outline: none;
}

.hamburger:not(.level1) .menuLink {
  padding: 0 0.3rem;
}

.title {
  display: inline-block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.icon {
  display: block;
  width: 1.25rem;
  min-width: 1.25rem;
  height: 1.25rem;
  background-position: center center;
  background-size: cover;
  background-repeat: no-repeat;
  margin: 0 0.5rem 0 0;
  -webkit-mask-size: cover;
  mask-size: cover;
  background-color: currentColor;
}

.label {
  top: 0;
  right: 0;
  display: inline-block;
  padding: 2px 3px;
  min-width: 42px;
  border-radius: 2px;
  background: var(--am-mega-menu-label-bg);
  color: var(--am-mega-menu-label-color);
  text-align: center;
  font-size: 9px;
  line-height: normal;
  transform: translateY(-50%);
  margin-left: 0.2rem;
}

.active {
  --venia-global-color-text: var(--am-mega-menu-active-color) !important;
}

/*
 * Desktop specific styles
 */

@media (min-width: 1025px) {
  .toggle {
    display: none;
    transform: none;
  }

  .toggle_open {
    transform: rotate(180deg);
  }

  .hamburger.level1 > .toggle {
    display: block;
  }

  .level2 > .menuLink {
    font-weight: 700;
    text-transform: uppercase;
    display: flex;
    white-space: normal;
  }

  .level2 .menuLink .title {
    white-space: normal;
  }

  .hamburger:not(.level1) .menuLink:hover {
    outline: 1px solid rgb(var(--am-mega-menu-text-hover-color));
  }

  .menuLink:hover {
    color: rgb(var(--am-mega-menu-text-hover-color));
    background-color: rgb(var(--am-mega-menu-item-bg-hover));
    outline: 0.5rem solid rgb(var(--am-mega-menu-item-bg-hover));
  }

  .hamburger.level1:hover > .toggle,
  .hamburger.level1:hover > .menuLink {
    background-color: rgb(var(--am-mega-menu-item-bg-hover));
  }

  .level2 .menuLink:hover {
    color: rgb(var(--am-mega-menu-submenu-text-hover-color));
    background: none;
    outline: none;
  }
}
