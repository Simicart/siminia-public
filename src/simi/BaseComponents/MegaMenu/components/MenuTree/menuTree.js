import React, { useMemo } from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { shape, string, array, bool } from 'prop-types';
import MenuItem from './menuItem';
import defaultClasses from './menuTree.module.css';

const MenuTree = props => {
  const { menuTree, config, view, classes: propsClasses, ...rest } = props;

  const menu = useMemo(
    () =>
      Array.isArray(menuTree) && menuTree.length
        ? menuTree.map(item => (
            <MenuItem key={item.id} view={view} {...item} {...rest} />
          ))
        : null,
    [menuTree, view, rest]
  );

  if (!menu) {
    return null;
  }

  const {
    ammegamenu_color_main_menu_background,
    ammegamenu_color_main_menu_background_hover,
    ammegamenu_color_main_menu_text,
    ammegamenu_color_main_menu_text_hover
  } = config;

  const style = {
    '--am-mega-menu-main-menu-background-color': ammegamenu_color_main_menu_background,
    '--am-mega-menu-main-menu-background-hover-color': ammegamenu_color_main_menu_background_hover,
    '--am-mega-menu-main-menu-text-color': ammegamenu_color_main_menu_text,
    '--am-mega-menu-main-menu-text-hover-color': ammegamenu_color_main_menu_text_hover,
}

  const classes = mergeClasses(defaultClasses, propsClasses);
  const menuClass = [classes.menu, classes[view]].join(' ');

  return (
    <nav id="menu-tree" className={classes.root} role="navigation" style={style}>
      <ul className={menuClass}>{menu}</ul>
    </nav>
  );
};

MenuTree.propTypes = {
  loading: bool,
  menuTree: array,
  view: string,
  classes: shape({
    root: string
  })
};

MenuTree.defaultProps = {
  menuTree: [],
  view: 'topMenu'
};

export default MenuTree;
