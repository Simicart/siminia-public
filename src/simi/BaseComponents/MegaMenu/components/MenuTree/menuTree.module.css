.root {
  align-items: center;
  align-self: center;
  grid-column: 3 / 4;
  grid-row: 1 / 1;
  justify-content: center;
  display: block;
  color: rgb(var(--am-mega-menu-main-menu-text-color));
  margin: 0;
  background-color: rgb(var(--am-mega-menu-main-menu-background-color));
  height: 100%;
}

.topMenu.menu {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  height: 100%;
}

@media(max-width: 1023px) {
  .root {
    padding: 10px 20px;
    color: rgb(var(--venia-global-color-text));
    background-color: #fff
  }
}