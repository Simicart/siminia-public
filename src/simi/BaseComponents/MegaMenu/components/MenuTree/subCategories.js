import React, { useMemo } from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { number, shape, string, array } from 'prop-types';
import MenuItem from './menuItem';

import defaultClasses from './subCategories.module.css';

const SubCategories = props => {
  const { items, level, column_count, view } = props;
  const subCategories = useMemo(() => {
    if (!Array.isArray(items) || !items.length) {
      return null;
    }

    return items.map(item => (
      <MenuItem key={item.id} {...item} view={view} level={level + 1} />
    ));
  }, [items, level, view]);

  if (!subCategories) {
    return null;
  }

  const classes = mergeClasses(defaultClasses, props.classes);
  const levelClass = classes[`level${level}`];
  const rootClass = [classes.root, levelClass, classes[view]].join(' ');

  return (
    <ul
      className={rootClass}
      style={{ '--am-mega-menu-columns-count': column_count || 4 }}
    >
      {subCategories}
    </ul>
  );
};

SubCategories.propTypes = {
  level: number,
  items: array,
  column_count: number,
  classes: shape({
    root: string
  })
};

SubCategories.defaultProps = {
  level: 0,
  items: [],
  column_count: 4
};

export default SubCategories;
