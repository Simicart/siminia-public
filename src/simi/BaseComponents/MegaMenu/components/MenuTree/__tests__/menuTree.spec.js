import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import MenuTree from '../menuTree';

jest.mock('../../MenuTree/menuItem', () => 'MenuItem');

test('renders without data', () => {
  const tree = createTestInstance(<MenuTree />);
  expect(tree.toJSON()).toMatchSnapshot();
});

const items = [
  {
    name: 'Bottoms',
    id: 'category-node-11',
    url: 'venia-bottoms.html',
    parent_id: null,
    __typename: 'MenuCategoryItem'
  },
  {
    name: 'Pants & Shorts',
    id: 'category-node-12',
    url: 'venia-bottoms/venia-pants.html',
    parent_id: 'category-node-11'
  }
];

test('renders correct tree', () => {
  const tree = createTestInstance(<MenuTree menuTree={items} />);
  expect(tree.toJSON()).toMatchSnapshot();
});
