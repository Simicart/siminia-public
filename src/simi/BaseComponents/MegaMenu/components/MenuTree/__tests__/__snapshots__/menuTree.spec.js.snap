// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct tree 1`] = `
<div>
  <ul
    className=" "
  >
    <MenuItem
      __typename="MenuCategoryItem"
      id="category-node-11"
      name="Bottoms"
      parent_id={null}
      url="venia-bottoms.html"
      view="topMenu"
    />
    <MenuItem
      id="category-node-12"
      name="Pants & Shorts"
      parent_id="category-node-11"
      url="venia-bottoms/venia-pants.html"
      view="topMenu"
    />
  </ul>
</div>
`;

exports[`renders without data 1`] = `null`;
